### ISSUE TITLE
Ask for meta data 

### ISSUE THREAD

**Issue by ZihuaZhao** [hhc1997/MSCN#2]
Hi hhc, I found that the meta data is needed to reproduce your experimental results. I wonder if you can provide us with your original meta data? Or any meta data splited from the training and validation sets is OK? Much thanks for your reply!

**Comment by hhc1997** [comment]
Hello, Zihua,
    Thank you for your interest in our work. 
    For the Flickr30K dataset, we randomly split the meta-data from the validation set:
`
    if opt.data_name == 'f30k_precomp':
        meta_len = 2900 # 2% of 145,000
        total_idsx = np.arange(0, len(images_meta)) #image length = caption length
        meta_idxs = np.random.choice(total_idsx, meta_len, False)
        captions_meta, images_meta = list(np.array(captions_dev)[meta_idxs]), images_dev[meta_idxs]
        #save...
`
  For the MS-COCO, the meta-data is split from the training set (6,328 pairs) and validation set (all 5,000 pairs):
`
    if opt.data_name == 'coco_precomp':
        im_div = [0, 1, 2, 3, 4]
        sup_len = 6328 # 2%*566435 - 5000
        total_img_idsx = np.arange(0, len(images_train))
        total_cap_idsx = np.arange(0, len(captions_train))
        sup_img_idxs = np.random.choice(total_img_idsx, sup_len, False)
        sup_0t4_idxs = np.random.choice(im_div, sup_len, True)
        sup_cap_idxs = sup_img_idxs * 5 + sup_0t4_idxs
        mask_img = np.ones(len(total_img_idsx), dtype=bool)
        mask_img[sup_img_idxs,] = False

        mask_cap = np.ones(len(total_cap_idsx), dtype=bool)
        del_cap_idxs = []
        for k in sup_img_idxs:
            del_cap_idxs.extend(list(range(k * len(im_div), k * len(im_div) + len(im_div))))
        del_cap_idxs = np.array(del_cap_idxs)
        mask_cap[del_cap_idxs,] = False
        ### get meta data
        img_meta_sup = images_train[sup_img_idxs]
        cap_meta_sup = list(np.array(captions_train)[sup_cap_idxs])
        images_meta = np.vstack((images_dev, img_meta_sup))
        captions_meta = captions_dev + cap_meta_sup
        ## get new train data
        images_train = images_train[mask_img]
        captions_train = list(np.array(captions_train)[mask_cap])
        #save    
`
For the CC152K, the meta-data is split from the validation set of the original Conceptual Captions. You can download the meta-data from https://drive.google.com/drive/folders/1XnGr7S-rXRfDbdeIF0QmTJV8kQFHx71-?usp=sharing.

**Comment by hhc1997** [comment]
Hello, @ZihuaZhao,
    Thank you for your interest in our work. 
    For the Flickr30K dataset, we randomly split the meta-data from the validation set:

```
if opt.data_name == 'f30k_precomp':
    meta_len = 2900 # 2% of 145,000
    total_idsx = np.arange(0, len(images_dev)) #image length = caption length
    meta_idxs = np.random.choice(total_idsx, meta_len, False)
    captions_meta, images_meta = list(np.array(captions_dev)[meta_idxs]), images_dev[meta_idxs]
    #save...
```

  For the MS-COCO, the meta-data is split from the training set (6,328 pairs) and validation set (all 5,000 pairs):

```
if opt.data_name == 'coco_precomp':
    im_div = [0, 1, 2, 3, 4]
    sup_len = 6328 # 2%*566,435 - 5000
    total_img_idsx = np.arange(0, len(images_train))
    total_cap_idsx = np.arange(0, len(captions_train))
    sup_img_idxs = np.random.choice(total_img_idsx, sup_len, False)
    sup_0t4_idxs = np.random.choice(im_div, sup_len, True)
    sup_cap_idxs = sup_img_idxs * 5 + sup_0t4_idxs
    mask_img = np.ones(len(total_img_idsx), dtype=bool)
    mask_img[sup_img_idxs,] = False

    mask_cap = np.ones(len(total_cap_idsx), dtype=bool)
    del_cap_idxs = []
    for k in sup_img_idxs:
        del_cap_idxs.extend(list(range(k * len(im_div), k * len(im_div) + len(im_div))))
    del_cap_idxs = np.array(del_cap_idxs)
    mask_cap[del_cap_idxs,] = False
    # get meta data
    img_meta_sup = images_train[sup_img_idxs]
    cap_meta_sup = list(np.array(captions_train)[sup_cap_idxs])
    images_meta = np.vstack((images_dev, img_meta_sup))
    captions_meta = captions_dev + cap_meta_sup
    # get new train data
    images_train = images_train[mask_img]
    captions_train = list(np.array(captions_train)[mask_cap])
    #save    
```

For the CC152K, the meta-data is split from the validation set of the original Conceptual Captions. You can download the meta-data from https://drive.google.com/drive/folders/1XnGr7S-rXRfDbdeIF0QmTJV8kQFHx71-?usp=sharing.

**Comment by ZihuaZhao** [comment]
Thank you for your quick response!
There's another question: I wonder if your are using the same correspondence noise index as NCR or DECL? Or you are using the simulated index randomly shuffled by yourself? I truely understand if there's a difference, but the experimental results can be better compared if the simulated datasets share a same noise index.

**Comment by hhc1997** [comment]
Hello, @ZihuaZhao,
I use the same noise index files provided by NCR (20% and 50%). For the noise index of 70% noisy correspondence, we employ the code provided in NCR (see line 58 in MSCN/data.py) to generate it.

**Comment by ZihuaZhao** [comment]
> Hello, @ZihuaZhao, Thank you for your interest in our work. For the Flickr30K dataset, we randomly split the meta-data from the validation set:
> 
> ```
> if opt.data_name == 'f30k_precomp':
>     meta_len = 2900 # 2% of 145,000
>     total_idsx = np.arange(0, len(images_meta)) #image length = caption length
>     meta_idxs = np.random.choice(total_idsx, meta_len, False)
>     captions_meta, images_meta = list(np.array(captions_dev)[meta_idxs]), images_dev[meta_idxs]
>     #save...
> ```
> 
> For the MS-COCO, the meta-data is split from the training set (6,328 pairs) and validation set (all 5,000 pairs):
> 
> ```
> if opt.data_name == 'coco_precomp':
>     im_div = [0, 1, 2, 3, 4]
>     sup_len = 6328 # 2%*566,435 - 5000
>     total_img_idsx = np.arange(0, len(images_train))
>     total_cap_idsx = np.arange(0, len(captions_train))
>     sup_img_idxs = np.random.choice(total_img_idsx, sup_len, False)
>     sup_0t4_idxs = np.random.choice(im_div, sup_len, True)
>     sup_cap_idxs = sup_img_idxs * 5 + sup_0t4_idxs
>     mask_img = np.ones(len(total_img_idsx), dtype=bool)
>     mask_img[sup_img_idxs,] = False
> 
>     mask_cap = np.ones(len(total_cap_idsx), dtype=bool)
>     del_cap_idxs = []
>     for k in sup_img_idxs:
>         del_cap_idxs.extend(list(range(k * len(im_div), k * len(im_div) + len(im_div))))
>     del_cap_idxs = np.array(del_cap_idxs)
>     mask_cap[del_cap_idxs,] = False
>     # get meta data
>     img_meta_sup = images_train[sup_img_idxs]
>     cap_meta_sup = list(np.array(captions_train)[sup_cap_idxs])
>     images_meta = np.vstack((images_dev, img_meta_sup))
>     captions_meta = captions_dev + cap_meta_sup
>     # get new train data
>     images_train = images_train[mask_img]
>     captions_train = list(np.array(captions_train)[mask_cap])
>     #save    
> ```
> 
> For the CC152K, the meta-data is split from the validation set of the original Conceptual Captions. You can download the meta-data from https://drive.google.com/drive/folders/1XnGr7S-rXRfDbdeIF0QmTJV8kQFHx71-?usp=sharing.

Thank you for your helpful responses, I've successfully generated the meta-data through your hints. There's something to mention that if "len(images_meta)" in the 3rd line for Flickr30K should be "len(images_dev)" according to your description?

By the way, I've downloaded the meta-data for CC152K, and found that the image data is stored in original images. NCR pre-processes the images into vectors through SCAN, which is found hard to implement since the processing code is based on Python2.7 and corresponding gcc version, which is not supported by new GPUs with new CUDA. I wonder if you have also processed the images through SCAN? Or you simply index in the processed vectors by NCR?

**Comment by hhc1997** [comment]
Hi，@ZihuaZhao. It should be `len(images_dev)` in the 3rd line for Flickr30K, and I have fixed this issue. For the meta-data in CC152K, I process the images through SCAN. You can download the processed features from https://drive.google.com/drive/folders/1XnGr7S-rXRfDbdeIF0QmTJV8kQFHx71-?usp=share_link

**Comment by ZihuaZhao** [comment]
Hi @hhc1997 , thank you a lot for providing so much details! 
I've found the original representations provided by SCAN is so limited, that further works can hardly be done. However the code under python2.7 and caffe can hardly be compiled. I've heard that you successfully extract the features using the SCAN code. It will be so kind if you can provide more details to run the code, including the conda environment or device version(3090 or A100 or something?) or your're using a docker? 
Thank you a lot for your replies again, and your answers do help.

**Comment by hhc1997** [comment]
Hi, @ZihuaZhao. Yes, compiling the SCAN code is hard on 3090 or A100. So I rent a server with a 2080Ti GPU on AutoDL (https://www.autodl.com/home).

**Comment by ZihuaZhao** [comment]
> Hi, @ZihuaZhao. Yes, compiling the SCAN code is hard on 3090 or A100. So I rent a server with a 2080Ti GPU on AutoDL (https://www.autodl.com/home).

Well, this is an available solution. Thank you :)

**Comment by yul1998** [comment]
> Hello, @ZihuaZhao, Thank you for your interest in our work. For the Flickr30K dataset, we randomly split the meta-data from the validation set:
> 
> ```
> if opt.data_name == 'f30k_precomp':
>     meta_len = 2900 # 2% of 145,000
>     total_idsx = np.arange(0, len(images_dev)) #image length = caption length
>     meta_idxs = np.random.choice(total_idsx, meta_len, False)
>     captions_meta, images_meta = list(np.array(captions_dev)[meta_idxs]), images_dev[meta_idxs]
>     #save...
> ```
> 
> For the MS-COCO, the meta-data is split from the training set (6,328 pairs) and validation set (all 5,000 pairs):
> 
> ```
> if opt.data_name == 'coco_precomp':
>     im_div = [0, 1, 2, 3, 4]
>     sup_len = 6328 # 2%*566,435 - 5000
>     total_img_idsx = np.arange(0, len(images_train))
>     total_cap_idsx = np.arange(0, len(captions_train))
>     sup_img_idxs = np.random.choice(total_img_idsx, sup_len, False)
>     sup_0t4_idxs = np.random.choice(im_div, sup_len, True)
>     sup_cap_idxs = sup_img_idxs * 5 + sup_0t4_idxs
>     mask_img = np.ones(len(total_img_idsx), dtype=bool)
>     mask_img[sup_img_idxs,] = False
> 
>     mask_cap = np.ones(len(total_cap_idsx), dtype=bool)
>     del_cap_idxs = []
>     for k in sup_img_idxs:
>         del_cap_idxs.extend(list(range(k * len(im_div), k * len(im_div) + len(im_div))))
>     del_cap_idxs = np.array(del_cap_idxs)
>     mask_cap[del_cap_idxs,] = False
>     # get meta data
>     img_meta_sup = images_train[sup_img_idxs]
>     cap_meta_sup = list(np.array(captions_train)[sup_cap_idxs])
>     images_meta = np.vstack((images_dev, img_meta_sup))
>     captions_meta = captions_dev + cap_meta_sup
>     # get new train data
>     images_train = images_train[mask_img]
>     captions_train = list(np.array(captions_train)[mask_cap])
>     #save    
> ```
> 
> For the CC152K, the meta-data is split from the validation set of the original Conceptual Captions. You can download the meta-data from https://drive.google.com/drive/folders/1XnGr7S-rXRfDbdeIF0QmTJV8kQFHx71-?usp=sharing.

Hi @hhc1997 @ZihuaZhao, may I ask where this code should be inserted?

**Comment by hhc1997** [comment]
> > Hello, @ZihuaZhao, Thank you for your interest in our work. For the Flickr30K dataset, we randomly split the meta-data from the validation set:
> > ```
> > if opt.data_name == 'f30k_precomp':
> >     meta_len = 2900 # 2% of 145,000
> >     total_idsx = np.arange(0, len(images_dev)) #image length = caption length
> >     meta_idxs = np.random.choice(total_idsx, meta_len, False)
> >     captions_meta, images_meta = list(np.array(captions_dev)[meta_idxs]), images_dev[meta_idxs]
> >     #save...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For the MS-COCO, the meta-data is split from the training set (6,328 pairs) and validation set (all 5,000 pairs):
> > ```
> > if opt.data_name == 'coco_precomp':
> >     im_div = [0, 1, 2, 3, 4]
> >     sup_len = 6328 # 2%*566,435 - 5000
> >     total_img_idsx = np.arange(0, len(images_train))
> >     total_cap_idsx = np.arange(0, len(captions_train))
> >     sup_img_idxs = np.random.choice(total_img_idsx, sup_len, False)
> >     sup_0t4_idxs = np.random.choice(im_div, sup_len, True)
> >     sup_cap_idxs = sup_img_idxs * 5 + sup_0t4_idxs
> >     mask_img = np.ones(len(total_img_idsx), dtype=bool)
> >     mask_img[sup_img_idxs,] = False
> > 
> >     mask_cap = np.ones(len(total_cap_idsx), dtype=bool)
> >     del_cap_idxs = []
> >     for k in sup_img_idxs:
> >         del_cap_idxs.extend(list(range(k * len(im_div), k * len(im_div) + len(im_div))))
> >     del_cap_idxs = np.array(del_cap_idxs)
> >     mask_cap[del_cap_idxs,] = False
> >     # get meta data
> >     img_meta_sup = images_train[sup_img_idxs]
> >     cap_meta_sup = list(np.array(captions_train)[sup_cap_idxs])
> >     images_meta = np.vstack((images_dev, img_meta_sup))
> >     captions_meta = captions_dev + cap_meta_sup
> >     # get new train data
> >     images_train = images_train[mask_img]
> >     captions_train = list(np.array(captions_train)[mask_cap])
> >     #save    
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For the CC152K, the meta-data is split from the validation set of the original Conceptual Captions. You can download the meta-data from https://drive.google.com/drive/folders/1XnGr7S-rXRfDbdeIF0QmTJV8kQFHx71-?usp=sharing.
> 
> Hi @hhc1997 @ZihuaZhao, may I ask where this code should be inserted?

Hi. You can run this code in a new Python file and save the 'images_meta ' and 'captions_meta ' as data_path/ data_name/meta_ims.npy' and 'data_path/data_name/meta_caps.txt', respectively.
